presto-main/src/main/java/com/facebook/presto/operator/DynamicFilterSourceOperator.java

```diff
@@ -0,0 +1,231 @@
+
+        private boolean closed;
+
+        public DynamicFilterSourceOperatorFactory(
+                int operatorId,
+                PlanNodeId planNodeId,
+                List<Type> types,
+                List<Integer> filterChannels,
+                Optional<Integer> hashChannel,
+                TupleDomainSource tupleDomainSource)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.types = ImmutableList.copyOf(types);
+            this.filterChannels = ImmutableList.<Integer>builder()
+                    .addAll(hashChannel.map(ImmutableList::of).orElse(ImmutableList.of()))
+                    .addAll(filterChannels)
+                    .build();
+            this.tupleDomainSource = requireNonNull(tupleDomainSource, "tupleDomainSource is null");
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return types;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterSourceOperator.class.getSimpleName());
+
+            return new DynamicFilterSourceOperator(operatorContext, types, filterChannels, tupleDomainSource);
+        }
+
+        @Override
+        public void close()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("Parallel dynamic filter build can not be duplicated");
+        }
+    }
+
+    private static final int DEFAULT_POSITIONS_LIMIT = 1000;
+
+    private final OperatorContext operatorContext;
+    private final List<Type> types;
+    private final List<Integer> filterChannels;
+    private final TupleDomainSource tupleDomainSource;
+    private final int positionsLimit;
+
+    private Page page;
+    private boolean finishing;
+    private long positionCount;
+    private List<ImmutableList.Builder<Object>> valuesBuilder;
+    private boolean[] nullsAllowed;
+
+    public DynamicFilterSourceOperator(
+            OperatorContext operatorContext,
+            List<Type> types,
+            List<Integer> filterChannels,
+            TupleDomainSource tupleDomainSource)
+    {
+        this(operatorContext, types, filterChannels, tupleDomainSource, DEFAULT_POSITIONS_LIMIT);
+    }
+
+    public DynamicFilterSourceOperator(
+            OperatorContext operatorContext,
+            List<Type> types,
+            List<Integer> filterChannels,
+            TupleDomainSource tupleDomainSource,
+            int positionsLimit)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.types = ImmutableList.copyOf(types);
```

Comment:
`ImmutableList.copyOf` checks for nullability already, so there's no need to check `requireNonNull` of its arguments.